### ISSUE TITLE
Tasks for Models

### ISSUE THREAD

**Issue by iamareebjamal** [iamareebjamal/open-event-android-sdk#6]
- [x] Add project structure and libraries
- [x] Convert Event API and model from JSON
- [x] Convert Microlocation API and model from JSON
- [x] Convert Track API and model from JSON
- [x] Convert Session Type API and model from JSON
- [x] Convert Sponsors API and model from JSON
- [x] Convert Speakers API and model from JSON
- [x] Convert Sessions API and model from JSON
- [x] Define and correctly convert Track relationships
- [x] Define and correctly convert Session relationships
- [x] Define and correctly convert Speaker relationships
- [x] Make event model consistent with old and new API
- [x] Make track model consistent with old and new API
- [x] Make speaker model consistent with old and new API
- [x] Make session model consistent with old and new API
- [x] Make microlocation model consistent with old and new API
- [x] Make sponsor model consistent with old and new API
- [x] Create Retrofit service classes to work with new API
- [x] Add orga models
- [ ] Add remaining links and meta without relationships

**Comment by iamareebjamal** [comment]
The last task is complex and will be broken into subtasks when we reach there

**Comment by iamareebjamal** [comment]
Currently working : 
~~@Shailesh351 - Session~~
~~@shreyasnbhat - Sponsors~~
~~@arp95 - Speaker~~

**Comment by iamareebjamal** [comment]
Let me refactor a little bit and then we'll work on adding relationships

**Comment by iamareebjamal** [comment]
I have refactored the project. Please take a look. I am currently working on Track Session relationship. Only Track to Session. Not the other way around

**Comment by arp95** [comment]
sure. then i am taking up the relationship between session to track

**Comment by Shailesh351** [comment]
I am taking up  session  relationship :+1:

**Comment by iamareebjamal** [comment]
IMO, whole session (to track and speaker) should be converted in one go. So please take up speaker to session relationship

**Comment by Shailesh351** [comment]
There is confusion regarding who is taking which part.

Currently working :
~@iamareebjamal : Track relationship
@Shailesh351 : Session relationship
@arp95 : speaker relationship~

fine ?

**Comment by iamareebjamal** [comment]
Yes

**Comment by shreyasnbhat** [comment]
@iamareebjamal  Any work for me?

**Comment by iamareebjamal** [comment]
@shreyasnbhat You can start adding meta and other relationship to Event class

**Comment by shreyasnbhat** [comment]
@iamareebjamal Will do and send a PR by tonight/tomorrow.

**Comment by iamareebjamal** [comment]
OK

**Comment by iamareebjamal** [comment]
OK guys, great job @arp95 @Shailesh351 Now we need to make old and new models as close as possible. So I'll be working on track and microlocation to trim out all the unnecessary annotations to make it look like the new model as much s we can. Remember to not remove any field, just annotations

**Comment by arp95** [comment]
so you are talking about the @JsonProperty() fields right? any other change?
i will work on speaker and sponser sections then.

**Comment by iamareebjamal** [comment]
Well our goal is to be able to merge new and old models, so whatever helps in that direction. And yes, I was talking about jsonproperty

**Comment by Shailesh351** [comment]
Sure. I am taking up other remaining models.
Along with it should we make old models keys according to new models ?

**Comment by iamareebjamal** [comment]
No, old model keys should not change. Or else the zip support will break. We'll analyze which keys are different and then report to Niranjan as he said 99% of time, same keys will be used

**Comment by arp95** [comment]
@iamareebjamal 
there are still classes where we can remove annotations. are you working on that or i can take up that?

**Comment by shreyasnbhat** [comment]
@iamareebjamal We don't add relationships to the event right? In all the models? Such as ApiSpeaker etc?

**Comment by arp95** [comment]
@shreyasnbhat 
in ApiSpeaker i added a relationship with session. i dont get what you exactly mean here?

**Comment by iamareebjamal** [comment]
Do you mean that the relationships in the event are not present yet? Yes, it's true but soft relationships and meta is there, so you should add them

**Comment by shreyasnbhat** [comment]
@arp95 What I mean is for example in the api/sponsors json file we have a relationship from sponsor to event. This is supposed to be added or not? As it hasn't been added and it creates a cyclic relationship.

**Comment by shreyasnbhat** [comment]
@iamareebjamal I also wanted to ask whether to include the Tax Info as we don't have a model for that and the next question is that is it needed?
And for the Links if we use the
`private Links links;` 
 we get something like this for the api/event.json `links='com.github.jasminb.jsonapi.Links@77cd7a0`

**Comment by iamareebjamal** [comment]
Well that's because it's an object. Don't add sponsor to event but we need meta and social links and other relationship link which were present in old models

**Comment by shreyasnbhat** [comment]
@iamareebjamal I'll add the Links as the jackson docs say and I will add all the relationships and meta in all the models?

**Comment by iamareebjamal** [comment]
All other models are complete in their requirements, just the Event one isn't

**Comment by iamareebjamal** [comment]
@arp95 Nice job on creating the comparison of APIs. Seems like Jackson is powerful enough to parse two different endpoint to same property https://stackoverflow.com/questions/19564711/json-jackson-parse-different-keys-into-same-field

So, even if the API have different end points, we'll make it work, but nevertheless, I'll discuss this

Furthermore, about the Event API, leave the fields which were added or removed, only the ones which were renamed. Can you include them in your doc. Thanks

Great work

**Comment by arp95** [comment]
@iamareebjamal sure. working on that.

**Comment by shreyasnbhat** [comment]
@iamareebjamal The Event already has the social links and I just have to add the relationship links right?

**Comment by iamareebjamal** [comment]
@shreyasnbhat The problem is that it can't be parsed using the new api as of now. You need to check out https://api.eventyay.com/#events-event-details and find how we can include social links in event and then update the json and parse it in the app so that we get all the information we could from the old json format

**Comment by Shailesh351** [comment]
We can get social links using `included` member

**Comment by arp95** [comment]
made changes:
Link: https://docs.google.com/document/d/1caaCeOOTIxQn2HR1Uj9r-cHXQH5hSYzzzCJpt3rkl38/edit

**Comment by shreyasnbhat** [comment]
@iamareebjamal What you are saying is the problem of not getting the social-links in one go right? We would have to do 2 network requests to get the social links. We get the link from the links for social-links and do another request. But the old model had the social links within the json itself. So we need the included section in the event json itself right? Am I on lines with what you are saying?

**Comment by iamareebjamal** [comment]
Correct

**Comment by shreyasnbhat** [comment]
@iamareebjamal  I'll just add the included section and continue adding the social-links

**Comment by arp95** [comment]
@iamareebjamal just to confirm in the speaker model we dont have "thumbnail" attribute?

**Comment by arp95** [comment]
Also we have taken everything as a string including the ids. is that intended because in the samples we are giving integer values.

**Comment by iamareebjamal** [comment]
It'll be converted to string automatically by Jackson. The new API has IDs in form of strings, that is why we have chosen string. However, I'll try some way to make it int

About thumbnail, nice catch. When models are merged, we will analyse the samples and add missing fields

**Comment by iamareebjamal** [comment]
~~@arp95 I checked and it is a requirement of json-api converter to use string IDs. No issues, we'll add migration in the app~~
We can do this

@Id(IntegerIdHandler.class)

**Comment by arp95** [comment]
sure as i am working on samples, i am having a look at all fields.

**Comment by iamareebjamal** [comment]
Also @Shailesh351 or @shreyasnbhat, one of you should open an issue on transfer from gson to jackson in Android Repo and start the required changes. Just so the app works with the current models without any error

**Comment by arp95** [comment]
I am updating the docs with the anomalies in attributes i am finding, so please have a look on that occasionally too.

**Comment by Shailesh351** [comment]
@iamareebjamal I don't understand why we need to transfer from gson to jackson. 
How is your plan to use this open-event-android-sdk  in android project ?

**Comment by iamareebjamal** [comment]
Well, everything we are doing here is through jackson, not gson

The plan was to create annotation processor, but now it won't be necessary

**Comment by arp95** [comment]
So now what is left over here? I mean we can move back to the main project now?

**Comment by iamareebjamal** [comment]
Yes you can move back to the main project till Niranjan responds with what to do with diverging fields

**Comment by shreyasnbhat** [comment]
@iamareebjamal  Should we make both the json files of the api and the older version same in terms of data for checking.

**Comment by iamareebjamal** [comment]
Sure, but don't invest too much time on it

**Comment by iamareebjamal** [comment]
@arp95 @Shailesh351 @shreyasnbhat Please start merging of models as discussed above whenever you get time

**Comment by arp95** [comment]
so if the attribute are different we should use the latest ones and add new attributes if there are any right?

**Comment by iamareebjamal** [comment]
The link I gave above gives us the ability to use same property for different attributes, so we won't need to delete anything, just add new methods and everything will work, even the old samples

https://stackoverflow.com/questions/19564711/json-jackson-parse-different-keys-into-same-field

**Comment by iamareebjamal** [comment]
Now everything is done, except some models from Orga Application, which I'll look into tomorrow. Now, we were supposed to create a library to share models, but since we need to extend models from different sources, like RealmModel in Open Event Android App and BaseModel in Orga App, I don't think it'll be possible. But next step would be to use a mock response from the server (like @arp95 made for the samples) and try to parse that with Retrofit using the jsonapi-converter. I'm a little busy for these 2 days because of festivities, but if you get time, please take a look at it

**Comment by iamareebjamal** [comment]
@Shailesh351 Please explain the bug you solved in list processing, I think it may be needed in the Open Event Android Parser too while parsing the static JSONs?

**Comment by Shailesh351** [comment]
@iamareebjamal Look at https://github.com/iamareebjamal/open-event-android-sdk/pull/19/files#diff-101323bdb4e416731a603a3a109aba5bL55

Previous implementation for List<T> was not correctly showing string(using toString() method). It was not actually parsing anything and  was just showing all the attributes from json file whether they are present or no in Model class. So i changed it.

**Comment by iamareebjamal** [comment]
Yeah, the toString was creating problems. but is there a need to add this fixes in Open Event Android App as well?

**Comment by Shailesh351** [comment]
~No it is not required.~ Sorry my bad. it is required to change it. We are using it in parsing jsons from assets


 And also i am having feeling that we can directly change network classes and models in open event android app and it will be sufficient for migrating to JSONAPI.

**Comment by arp95** [comment]
@Shailesh351 well that is possible but the whole point of this project is to maintain a library which can be used in both projects. 
Changing the models according to the api is an obvious approach i think.

**Comment by shreyasnbhat** [comment]
@iamareebjamal How will using this as a library work exactly?

**Comment by iamareebjamal** [comment]
I was talking about the list parsing fixes @Shailesh351 made in this project, should they be added to opev android as well. About this library @arp95, it is not possible to use models from here because both projects have different base classes. So we can't make them share models

**Comment by arp95** [comment]
@iamareebjamal so should I work on the open-event android project and change the models according t the new api?

**Comment by iamareebjamal** [comment]
Sure, try to conform to the models we have created here as much as possible. And remember that we don't need to change field names, just add appropriate setters with annotations. If you add fields, please add corresponding Realm Migrations, so I'll recommend to break it in smaller PRs

**Comment by rohanagarwal94** [comment]
@iamareebjamal was busy with the livefeed issue. The orga models are still left right? I have started working on them.

**Comment by iamareebjamal** [comment]
I added the ones which are available in the docs in #22

**Comment by rohanagarwal94** [comment]
Okay so anything left for the models?

**Comment by iamareebjamal** [comment]
You can start adding type and relationship attributes to opev android models

**Comment by rohanagarwal94** [comment]
@iamareebjamal on it.

**Comment by rohanagarwal94** [comment]
@iamareebjamal as far as I see, the relationships are defined in the open event android models. Are there any other relationships which are not defined?

**Comment by arp95** [comment]
there needs to be added new attributes which arent there in the android part yet. take up models and open an issue there.
i am working on new attributes for session. please check the response for others and work. remember you have to modify the realm part as well.

**Comment by rohanagarwal94** [comment]
Apart from sessions, has someone started adding new attributes to any other model? Please confirm before I open an issue for speakers.

**Comment by iamareebjamal** [comment]
I don't think anyone else is working on any other model (y)

**Comment by iamareebjamal** [comment]
@arp95 @Shailesh351 @rohanagarwal94 Please review the PRs and change the models in Opev Android quickly as the API is now live (except speaker part)

**Comment by arp95** [comment]
sure areeb. also can we start removing the old attributes?
i think as such only event model is left which i am works ng on.

**Comment by Shailesh351** [comment]
I am working on adding Type, Id and relationship annotations in models. Will send PR by today
